src/audio/rtnr/rtnr.c

```diff
@@ -594,23 +652,30 @@ static int rtnr_prepare(struct comp_dev *dev)
 	if (ret == COMP_STATUS_STATE_ALREADY_SET)
 		return PPL_STATUS_PATH_STOP;
 
-	/* Get sink data format */
-	sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
-	cd->sink_format = sinkb->stream.frame_fmt;
+	/* TODO(Realtek): move the config fetch codes here */
 
-	/* Check source and sink PCM format and get copy function */
-	comp_info(dev, "rtnr_prepare(), sink_format=%d", cd->sink_format);
-	cd->rtnr_func = rtnr_find_func(cd->sink_format);
-	if (!cd->rtnr_func) {
-		comp_err(dev, "rtnr_prepare(): No suitable processing function found.");
-		goto err;
-	}
+	if (cd->process_enable) {
+		/* Get sink data format */
+		sinkb = list_first_item(&dev->bsink_list, struct comp_buffer, source_list);
+		cd->sink_format = sinkb->stream.frame_fmt;
 
-	/* Default on */
-	cd->process_enable = true;
+		/* Check source and sink PCM format and get copy function */
+		comp_info(dev, "rtnr_prepare(), sink_format=%d", cd->sink_format);
+		cd->rtnr_copy_func = rtnr_process;
+		cd->rtnr_func = rtnr_find_func(cd->sink_format);
+		if (!cd->rtnr_func) {
+			comp_err(dev, "rtnr_prepare(): No suitable processing function found.");
+			goto err;
+		}
 
-	/* Clear in/out buffers */
-	RTKMA_API_Prepare(cd->rtk_agl);
+		/* Clear in/out buffers */
+		RTKMA_API_Prepare(cd->rtk_agl);
+	} else {
+		/* Bypassthrough path is selected */
+		comp_info(dev, "rtnr_prepare(), passthrough mode");
+		cd->rtnr_copy_func = rtnr_passthrough;
+		cd->rtnr_func = NULL;
+	}
```

Comment:
@johnylin76 Thanks for the PR! 
@cujomalainey Do we have to support toggling on/off on-the-fly during recording? If so I think we should move this part to rtnr_set_value() since copy function is assigned before recording and couldn't be changed on-the-fly.

Reply:
Toggling on/off on-the-fly during recording should be supported. The fact is that it will be handled on CRAS side.
The input device node (e.g. iternal microphone) will be closed and then re-opened transiently while NR state is toggled.

In the aspect of SOF side, RTNR, when the state is being toggled, it will be triggered to stop (comp_reset) and then start (comp_prepare, comp_trigger). So, it is sufficient to make the config and state update just during comp_prepare.

And here I mean is for the toggle of the real use cases, not including the toggle by amixer command because it will not be across CRAS.

Reply:
Thanks for the description. For developers like us this might be helpful for debugging. Would there be a chance that sof+RTNR is running on the device without NS UI integrated in the OS?

Reply:
Some components would add a config-changed check in comp_copy(), for example:
https://github.com/thesofproject/sof/blob/main/src/audio/eq_iir/eq_iir.c#L825
which makes it able to "dynamically" update config on-the-fly. This is for config blob

I am not sure for switch-typed value change, with the concern that whether the setter (comp_cmd) and the getter (comp_copy) are synchronized. The former is from IPC while the latter is from scheduler. To be honest I am not sure whether the config blob had handled synchronization either.
@singalsu hi Seppo could you help to comment on this?

Reply:
There's no controls synchronization in the framework from cmd() to copy(). If a control could create a hazard if it changes in the middle of processing the component should implement some two phase control change. The binary blob is updated in two steps by the model handler so it doesn't need work in the component side. Expect if @jsarha 's new single blob feature (for very large ones) is used.